proposals/NNNN-DXIL-Scalarization.md

```diff
@@ -0,0 +1,239 @@
+"HLSL vectors are scalarized" and "Matrices are lowered to vectors". Therefore,
+ we need to be able to scalarize these types in clang DirectX backend. In DXC, 
+ this is done via [`SROA_Parameter_HLSL` Module pass in `ScalarReplAggregatesHLSL.cpp`](https://github.com/microsoft/DirectXShaderCompiler/blob/main/lib/Transforms/Scalar/ScalarReplAggregatesHLSL.cpp#L4263). 
+ The goal of this proposal is the present a solution for LLVM's pass manager.
+
+## Motivation
+Without Scalarizing the data structures and call instructions we can't generate valid DXIL.
+
+ ## Background
+The rewrite functions specified in the appendix cover the behavior we need to 
+support for scalarization. The specific instructions DXC operates on are
+`CallInst`, `LoadInst`, `StoreInst`, `GetElementPtrInst`, `AddrSpaceCastInst`, `BitCastInst`, 
+and `MemIntrinsic`. DXC  works on two sets of data globals\allocas and 
+function arguments via the `RewriteForScalarRepl`.
+
+There are a few ways we could tackle this work that gets us towards behavioral parity.
+A good example is to look at:
+- [SPIRVEmitIntrinsics.cpp](https://github.com/llvm/llvm-project/blob/main/llvm/lib/Target/SPIRV/SPIRVEmitIntrinsics.cpp)
+- [SROA.cpp](https://github.com/llvm/llvm-project/blob/main/llvm/lib/Transforms/Scalar/SROA.cpp)
+- [Scalarizer.cpp](https://github.com/llvm/llvm-project/blob/main/llvm/lib/Transforms/Scalar/Scalarizer.cpp)
+
+Given how the existing passes are written we should get `GetElementPtrInst`, `AddrSpaceCastInst`, `BitCastInst`, 
+and `MemIntrinsic` support mostly for free by tapping into functions like 
+`visitGetElementPtrInst`, `visitBitCastInst`,  `visitMemSetInst`, & `visitMemTransferInst`.
+
+There likely shouldn't be any need for `AddrSpaceCastInst`. 
+
+We also will get `LoadInst` and `StoreInst` for non DXIL resources  via  `visitLoadInst` and `visitStoreInst`.
+
+Finally the `CallInst` cases is really just to cover HLSL intrinsics. This can be handled via a simple pass
+that knows all the intrinsics and loops through the vector type to emit the correct number of calls per element.
+
+## Priority of work
+- A good starting place of behaviors to try to match is `RewriteCall` 
+specifically `RewriteWithFlattenedHLIntrinsicCall`As the other pieces are not needed 
+for compute shaders. 
+- Next would be `RewriteForLoad` and `RewriteForStore` as those 
+transformations are need to properly handle resources and breakdown to 
+three simple cases.
+
+The behaviors of `RewriteForGEP`, `RewriteForAddrSpaceCast`, `RewriteForConstExpr`,
+and `RewriteMemIntrin`. Should be things we can get  via existing LLVM passes.
+
+## Areas explicitly ignored
+SROA_Parameter pass has some  matrix loads and stores that this proposal won't go into.
+It also looks to do some legalization via `LegalizeDxilInputOutputs` that won't be covered.
+
+## Tickets
+1. Create a scalarization pass that just handles the flattenArgument case to start.      
```

Comment:
I'd like to understand this more. Do you have an example before and after IR for a rewritten call instruction?

Reply:
Yeah I have two approaches one thats easy to move and one thats less easy to move around. In both cases I need some indication that we want to scalarize. I was looking into how to use the `LowerMap` in `DXILOperation.inc`. But even then I would need a flag because some intrinsics like dot2-4  shouldn't be scalarized.   `DXIL.td` will have a bool  we set on whether to scalarize or not scalarize an intrinsic.

- case 1 the pass we can't move
We modify  `lowerIntrinsic` in `llvm/lib/Target/DirectX/DXILOpLowering.cpp` to check for this flag set or not then directly replace an intrinsics with a set of DXILOps.  In this case we do scalarization as part of DXIL lowering. Advantage to doing it this way is we are limiting the intrinsics to just those that map to DXIL Ops which means we don't have to check intrinsic scalarization across all CallInst.

- case 2 we replace 1 intrinsic with more intrinsics
Doing it this way we can put this pass any where before DXIL Lowering. And the function to do this is very simple:
```cpp
static bool scalarizeIntrinsic(Function &F, CallInst *OrigIntrinsic) {
  Instruction *CurrInstruction = OrigIntrinsic;
  IRBuilder<> Builder(CurrInstruction);

  size_t NumOperands = OrigIntrinsic->getNumOperands();
  Value *Arg0 = OrigIntrinsic->getOperand(0);
  auto *VecArg = dyn_cast<FixedVectorType>(Arg0->getType());
  Value *Result = llvm::PoisonValue::get(llvm::FixedVectorType::get(
      VecArg->getElementType(), VecArg->getNumElements()));
  for (size_t i = 0; i < VecArg->getNumElements(); i++) {
    SmallVector<Value *, 4> ExtractedElements;
    for (size_t j = 0; j < NumOperands; j++) {
      Value *CurrArg = OrigIntrinsic->getOperand(j);
      if (CurrArg->getType()->isPointerTy())
        continue;
      Value *ArgIndex =
          ConstantInt::get(Type::getInt32Ty(CurrArg->getContext()), i);
      Value *ExtractedElement = Builder.CreateExtractElement(CurrArg, ArgIndex);
      ExtractedElements.push_back(ExtractedElement);
    }
    ArrayRef<Value *> ArgsAsArrayRef(ExtractedElements.data(),
                                     ExtractedElements.size());
    CallInst *CurrCall = Builder.CreateIntrinsic(
        /*ReturnType*/ VecArg->getElementType(), F.getIntrinsicID(),
        ArgsAsArrayRef);
    Result = Builder.CreateInsertElement(Result, CurrCall, i);
  }

  OrigIntrinsic->replaceAllUsesWith(Result);
  OrigIntrinsic->eraseFromParent();
  return true;
}
```

Reply:
- example for  lowerintrinsic
```llvm
; CHECK:extractelement <{{.*}} x float> %{{.*}}, i32 {{.*}}
; CHECK:call float @dx.op.unary.f32(i32 13, float %{{.*}})
; CHECK:insertelement <{{.*}} x float>
; CHECK:extractelement <{{.*}} x float> %{{.*}}, i32 {{.*}}
; CHECK:call float @dx.op.unary.f32(i32 13, float %{{.*}})
; CHECK:insertelement <{{.*}} x float>
 %elt.sin = call <2 x float> @llvm.sin.v2f32(<2 x float> %0)
```
-example as seperate pass
```llvm
; CHECK:extractelement <{{.*}} x float> %{{.*}}, i32 {{.*}}
; CHECK:call float @llvm.sin.f32(float %{{.*}})
; CHECK:insertelement <{{.*}} x float>
; CHECK:extractelement <{{.*}} x float> %{{.*}}, i32 {{.*}}
; CHECK:call float @llvm.sin.f32(float %{{.*}})
; CHECK:insertelement <{{.*}} x float>
 %elt.sin = call <2 x float> @llvm.sin.v2f32(<2 x float> %0)
```

Reply:
Oaky, that's fine for an element wise operation, but what about something like `dot`?

Reply:
I don't think I understand why we need the "flexibility" to move this around. It seems to me that we're unlikely to want to run a lot of optimizations between scalarizing intrinsic calls and generating DXIL since the point of preserving the vectors is to allow LLVM to optimize more effectively.

Reply:
> Oaky, that's fine for an element wise operation, but what about something like `dot`?

We did dot already:
https://github.com/llvm/llvm-project/blob/ea8bb4d633683f5cbfd82491620be3056f347a02/llvm/lib/Target/DirectX/DXILOpLowering.cpp#L86-L89
In this PR:
https://github.com/llvm/llvm-project/pull/86071/files#diff-abc4769e41f4d9022e3b88d8d752267e5d1bb26c0d4c4186f5d551c0dce76b80

Reply:
> I don't think I understand why we need the "flexibility" to move this around. It seems to me that we're unlikely to want to run a lot of optimizations between scalarizing intrinsic calls and generating DXIL since the point of preserving the vectors is to allow LLVM to optimize more effectively.

Maybe we don't I'll confirm the use case i'm worried about with you offline.